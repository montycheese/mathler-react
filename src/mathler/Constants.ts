export enum MathlerTileState {
    CORRECT_PLACE = 0, // green
    DIFFERENT_PLACE = 1,// orange
    INCORRECT_VALUE = 2, // gray
    NO_VALUE = 3, // blank
    PENDING_SUBMISSION = 4 // white
};

export const VALID_CHARS = '1234567890+-/*';
export const SUPPORTED_OPERATORS = '+-/*';
export const MAX_CHARS_PER_ROW = 6;
export const MAX_TILES = 36;
