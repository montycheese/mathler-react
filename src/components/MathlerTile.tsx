import {MathlerTileState} from "../mathler/Constants";
import {useEffect, useRef} from "react";
import './MathlerTile.css'

type MathlerTileProps = {
    input: string;
    state: MathlerTileState
}

export default function MathlerTile({ input, state }: MathlerTileProps) {

    const prevTileStateRef = useRef<MathlerTileState>(MathlerTileState.NO_VALUE);

    useEffect(() => {
        prevTileStateRef.current = state;
    }, [state]);

    let bkgColorClass = 'bg-white';
    let animationClasses = '';
    switch (state) {
        case MathlerTileState.CORRECT_PLACE:
            bkgColorClass = 'bg-green-500';
            if (prevTileStateRef.current !== MathlerTileState.CORRECT_PLACE) {
                animationClasses += 'zoom';
            }
            break;
        case MathlerTileState.DIFFERENT_PLACE:
            bkgColorClass = 'bg-orange-500';
            break;
        case MathlerTileState.INCORRECT_VALUE:
            bkgColorClass = 'bg-slate-400';
            break;
    }

    return (
        <div className={`${bkgColorClass} ${animationClasses} rounded-md text-black grid place-items-center p-2 w-10 h-10 hover:bg-cyan-100`}>
            {input}
        </div>
    );
}

MathlerTile.getBlankTile = (key: string) => <MathlerTile input="" state={MathlerTileState.NO_VALUE} key={key} />;
