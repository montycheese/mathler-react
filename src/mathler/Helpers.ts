import {SUPPORTED_OPERATORS} from "./Constants";

export function isValidEquation(equation: string): boolean {
    let isValid = false;

    // ensure that the equation contains at least one operator.
    for (let i = 0; i < SUPPORTED_OPERATORS.length; i++) {
        if (equation.includes(SUPPORTED_OPERATORS[i])) {
            isValid = true;
            break;
        }
    }

    try {
        const result = eval(equation);
        isValid = isValid && typeof result === 'number' && result >= 0;
    } catch (error) {
        console.error(error);
    }
    return isValid;
}
