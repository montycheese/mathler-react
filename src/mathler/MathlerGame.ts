// represents a game instance of Mathler
import {MathlerTileState} from "./Constants";

export default class MathlerGame {
    public calculation: string;
    public value: number;
    public submissions: string[][];
    public isGameOver: boolean;

    constructor(calculation: string, value: number) {
        this.calculation = calculation;
        this.value = value;
        //const submissions = [];
       // for (let i = 0; i < 6; i++) {
       //     submissions.push(Array(6).fill(''))
       // }
        this.submissions = []; //submissions;
        this.isGameOver = false;
    }

    submitNewRow(submission: string[]) {
        if (this.submissions.length < 6) {
            this.submissions.push(submission);
            if (this.isValidEquation(submission.join(''))) {
                this.isGameOver = true;
            }
        }
    }

    isValidEquation(calculation: string): boolean {
        return eval(calculation) == this.value
            && this.checkSameCharacters(calculation, this.calculation);
    }

    getCharState(char: string, index: number): MathlerTileState {
        if (char.length !== 1) {
            throw new Error('Invalid char. Must be of length 1.');
        }
        if (this.calculation[index] === char) {
            return MathlerTileState.CORRECT_PLACE;
        } else if (this.calculation.includes(char)) {
            return MathlerTileState.DIFFERENT_PLACE;
        }
        return MathlerTileState.INCORRECT_VALUE;
    }

    checkSameCharacters(str1: string, str2: string) {
        // Check if the lengths of the cleaned strings are equal
        if (str1.length !== str2.length) {
            return false;
        }

        // Create frequency maps for each string
        const frequencyMap1 = {};
        const frequencyMap2 = {};

        // Populate frequency maps for str1
        for (let char of str1) {
            // @ts-ignore
            frequencyMap1[char] = (frequencyMap1[char] || 0) + 1;
        }

        // Populate frequency maps for str2
        for (let char of str2) {
            // @ts-ignore
            frequencyMap2[char] = (frequencyMap2[char] || 0) + 1;
        }

        // Check if the frequency maps are equal
        for (let char in frequencyMap1) {
            // @ts-ignore
            if (frequencyMap1[char] !== frequencyMap2[char]) {
                return false;
            }
        }

        return true;
    }

}
