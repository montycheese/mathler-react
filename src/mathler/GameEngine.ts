import MathlerGame from "./MathlerGame";

const puzzles = [
    '24*2-9',
    '27*3-9',
    '95/5+8',
    '1+2+33'
];

export default class GameEngine {
    public currentInstance: MathlerGame | null;

    constructor() {
        this.currentInstance = null;
    }

    getGameInstance = (): MathlerGame => {
        return this.currentInstance || this.startNewGameInstance();
    };

    startNewGameInstance = (): MathlerGame => {
        const equation = puzzles[Math.floor(Math.random() * puzzles.length)];
        const value = eval(equation);
        console.log(`Starting new game instance with eq: [${equation}] and value: [${value}]`);
        this.currentInstance = new MathlerGame(equation, value);
        return this.currentInstance;
    }

}
