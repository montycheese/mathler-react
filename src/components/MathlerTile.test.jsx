import React from 'react';
import renderer from 'react-test-renderer';
import MathlerTile from "./MathlerTile";
import {MathlerTileState} from "../mathler/Constants";

it('renders "no value" state correctly', () => {
    const component = renderer.create(
        <MathlerTile
            input={''}
            state={MathlerTileState.NO_VALUE}
        />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders "incorrect value" state correctly', () => {
    const component = renderer.create(
        <MathlerTile
            input={'1'}
            state={MathlerTileState.INCORRECT_VALUE}
        />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders "different place value" state correctly', () => {
    const component = renderer.create(
        <MathlerTile
            input={'1'}
            state={MathlerTileState.DIFFERENT_PLACE}
        />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders "correct place value" state correctly', () => {
    const component = renderer.create(
        <MathlerTile
            input={'1'}
            state={MathlerTileState.CORRECT_PLACE}
        />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders "pending submission value" state correctly', () => {
    const component = renderer.create(
        <MathlerTile
            input={'+'}
            state={MathlerTileState.PENDING_SUBMISSION}
        />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

