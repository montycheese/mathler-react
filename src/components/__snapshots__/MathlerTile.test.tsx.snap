// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders "correct place value" state correctly 1`] = `
<div
  className="bg-green-500 rounded-md text-black grid place-items-center p-2 w-10 h-10 hover:bg-cyan-100"
>
  1
</div>
`;

exports[`renders "different place value" state correctly 1`] = `
<div
  className="bg-orange-500 rounded-md text-black grid place-items-center p-2 w-10 h-10 hover:bg-cyan-100"
>
  1
</div>
`;

exports[`renders "incorrect value" state correctly 1`] = `
<div
  className="bg-slate-400 rounded-md text-black grid place-items-center p-2 w-10 h-10 hover:bg-cyan-100"
>
  1
</div>
`;

exports[`renders "no value" state correctly 1`] = `
<div
  className="bg-white rounded-md text-black grid place-items-center p-2 w-10 h-10 hover:bg-cyan-100"
/>
`;

exports[`renders "pending submission value" state correctly 1`] = `
<div
  className="bg-white rounded-md text-black grid place-items-center p-2 w-10 h-10 hover:bg-cyan-100"
>
  +
</div>
`;
